metrics/src/agent.rs

```diff
@@ -0,0 +1,267 @@
+
+use hyper::header::CONTENT_TYPE;
+use hyper::http::HeaderValue;
+use hyper::{Body, Client, Request};
+use hyper_rustls::HttpsConnectorBuilder;
+
+const DOGSTATSD_PORT: u16 = 8125;
+const BUFFER_SIZE: usize = 8192;
+
+pub struct MetricsAgent {
+    config: Config,
+    buf: Arc<Mutex<Vec<Metric>>>,
+}
+
+impl MetricsAgent {
+    pub fn with_config(config: Config) -> Self {
+        Self {
+            config,
+            buf: Arc::new(Mutex::new(Vec::new())),
+        }
+    }
+
+    pub async fn run(&self) {
+        // Create a UDP socket and bind to the dogstatsd port
+        let socket = UdpSocket::bind(SocketAddr::new(
+            std::net::Ipv4Addr::new(0, 0, 0, 0).into(),
+            DOGSTATSD_PORT,
+        ))
+        .await
+        .expect("Error binding to socket");
+
+        println!("Listening for dogstatsd packets on port {}", DOGSTATSD_PORT);
+
+        // We use a shared vec here instead of a channel because we need to buffer the data at
+        // some point anyways. If we had used a channel here, we'd still need to add it to
+        // some type of buffer, so we'd need to implement a vector buffer anyways.
+        // Instead, we can save a step and just add it to the buffer directly.
+        let buf_producer = self.buf.clone();
+        let buf_consumer = self.buf.clone();
+
+        let https = HttpsConnectorBuilder::new()
+            .with_native_roots()
+            .https_only()
+            .enable_http1()
+            .build();
+        let http_client = Client::builder().build::<_, Body>(https);
+
+        // Process DogstatsD UDP packets and write them to our shared buffer
+        tokio::spawn(async move {
+            loop {
+                let mut buffer = [0; BUFFER_SIZE];
+                let bytes_received = match socket.recv_from(&mut buffer).await {
+                    Ok((size, _)) => size,
+                    Err(e) => {
+                        println!("Error receiving bytes from UDP packet: {}", e);
+                        continue;
+                    }
+                };
+
+                let unix_timestamp = match SystemTime::now().duration_since(UNIX_EPOCH) {
+                    Ok(duration) => duration.as_secs(),
+                    Err(e) => {
+                        println!("Error generating unix timestamp: {}", e);
+                        continue;
+                    }
+                };
+
+                let metrics: Vec<Metric> = match str::from_utf8(&buffer[..bytes_received]) {
+                    Ok(metrics_str) => metrics_str
+                        .split('\n')
+                        .filter_map(|metric_str| Metric::from_string(metric_str, unix_timestamp))
+                        .collect(),
+                    Err(e) => {
+                        println!("Error converting metric to str: {}", e);
+                        continue;
+                    }
+                };
+
+                let mut metrics_to_process = buf_producer.lock().await;
```

Comment:
Yeah! So I realized that we'd still need a buffer to store the metrics until we flush them to Datadog. If we used a channel, we'd still need to construct this buffer, and write to it. We'd essentially have the exact same problem (shared memory), so instead I opted to just write to the buffer directly instead of using a channel. This is because we need to flush every n seconds, and we need a place to store the Metrics until we flush. For example:

```rust
let (tx, rx) = mspc::channel();
let buf Vec<Metric> = vec![];

loop {
  let metric = get_metric_from_udp_server();
  tx.send(metric);
}

loop {
  rx.recv(metric);
  buf.push(metric);
  
  // what happens here? if we want to flush every 3s (buffered, we don't just want to 
  // flush single metrics), we need a new thread,
  // which means we need to put buf behind an Arc<Mutex<>>
  // which means we have the same problem as what we have now
}
```

I'm not sure if there's a better way to solve this problem, so I opted for the current approach. Do you have any suggestions?

Reply:
You could do something like this with channels:
 ```rust
let mut buffer = vec![]; // Reuse the same vec every iteration to avoid allocations
loop {
    buffer.clear();
    if let Some(iter) = rx.try_iter() {
        buffer.extend(iter);
    }
}
```
I'm not sure though. Do you think that's any neater?

Reply:
I think the main difference is this won't lock either thread. It's possible the producer thread will be blocked from writing while the lock is held by the consumer, which might be as long as it takes to flush the metrics to the backend.

Reply:
I've gone ahead and implemented this.

Looking back, I'm not entirely sure why I thought we'd need another thread poll the buf every 3s. This approach works, and is a lot simpler, much appreciated!